### PR TITLE
Making `icu_locid_macros` into `consts`

### LINKED ISSUE #348: Re-write locale macros as const functions
Proc macros are clunky and error-prone.  Problems include:

1. You cannot depend on `icu_locid_macros` by itself; you need to include either `icu` or `icu_locid`
2. They are heavier than `macro_rules!` and `const fn`, since the compiler needs to link and run them as a separate binary
3. The macros crate needs to be published separately on crates.io
4. Pulls in a nontrivial `proc-macro-crate` dependency in Cargo.toml

We are close to being able to write the locale macros as const functions.

I previous said in https://github.com/unicode-org/icu4x/issues/310#issuecomment-707494987:

> Okay, I also do not see any sign of string/slice indexing operations. I saw it suggested in [rust-lang/rust#52000](https://github.com/rust-lang/rust/issues/52000) that this could change (allowing `for` loops), but there doesn't seem to be a lot of movement on the idea of the fundamental indexing operation in const functions.
> 
> That being said, since we can pretty trivially implement things like `TinyStr4::is_ascii_alphabetic()` as const functions, we could limit the locale macro to be along the lines of
> 
> ```
> langid!("en")
> langid!("en", "US")
> langid!("en", "Latn", "US")
> ```
> 
> We can validate the subtags, but we can't parse the `"en-Latn-US"` syntax inside the const function in any clean way. (I guess you could convert the string to a TinyStr16 and do bitwise operations on that...)

I think we should consider this route as a more future-proof solution to this problem.

### PR DESCRIPTION
**Pull request by robertbastian**
Fixes #348

### CODE REVIEW

**Review by sffc — DISMISSED** [review]
Looks good from my end!  Please wait for @zbraniecki's review.

File: components/locid/src/langid.rs
Comment (zbraniecki) [inline components/locid/src/langid.rs:93]: `suggestion`: the comment would be more actionable if it was phrased as "Switch to `Result<Self, ...>` once X stabilizes. See ...`

File: components/locid/src/parser/mod.rs
Comment (zbraniecki) [inline components/locid/src/parser/mod.rs:36]: `suggestion`: `t` is a very cryptic name for a field that is carried around. Can you call it `slice` or `source`? For `l_cursor` and `r_cursor` - consider `start`/`end`. or `start_idx` and `end_idx`.

File: components/locid/src/subtags/region.rs
Comment (zbraniecki) [inline components/locid/src/subtags/region.rs:58]: `question`: this should cost perf. You are forcing tinystr construction for each attempt to test if the tag is a region. Before that, the code is heavily optimized to bail with as few operations as possible ordered from the most likely reasons to bail to the least likely.
This allows consumers to write all their code with `if let Ok(region) = Region::try_from_x()` quite abundantly.

I think that since multiple subtags can match the same string, it is a good practice to keep it, over separating `does_slice_match` vs `from_slice` for testing/constructing.

Can you look back at the from slice constructors that you rewrote and shift from optimizing for code conscience to optimize for cheap bailouts and nest more to only pay the cost when needed? (e.g. testing length is cheap, so can be done early, but construction of tinystr is not, so do it only if length matches etc.).

**Review by zbraniecki — COMMENTED** [review]
overall it looks great! Thank you!

I am sad that const limitations make us have to sacrifice readability for const compatibility, but I believe it is a right tradeoff.
We can revisit and refactor the code when Rust const ergonomics are complete (`?` would help a lot).

I left a couple notes, but my two overarching ones are:
1) Please, do not forgo the value of runtime parsing. icu_locid should be optimized for scenarios where code receives a string and needs to parse it at runtime. It's a very very hot loop.
2) It would be great to find a good place to document the fact that we're giving up on readability to get consts and that this code should be refactored back once Rust const ergonomics are in place. Maybe as a comment over `LanguageIdentifier` main parser function?

File: components/locid/src/subtags/region.rs
Comment (zbraniecki) [reply components/locid/src/subtags/region.rs:58]: (I am making an assumption above that rust match for tuples is not smart enough to delay second item in the tuple until first one matches, but I didn't check the assembly. Feel free to evaluate)

**Comment by robertbastian** [comment]
> It would be great to find a good place to document the fact that we're giving up on readability to get consts and that this code should be refactored back once Rust const ergonomics are in place. Maybe as a comment over LanguageIdentifier main parser function?

We haven't done this in other places either, I'm thinking we can assume that if it's a const function it's clear that we don't get `?`, for loops, etc.

File: components/locid/src/subtags/language.rs
Comment (zbraniecki) [inline components/locid/src/subtags/language.rs:82]: `question`: am I correct that this will not work for non-canonicalized case for `und`? Like `UND` will not return `None` here?

Comment (zbraniecki) [reply components/locid/src/subtags/language.rs:82]: I think this is a pre-existing bug, but may be worth fixing here.

**Review by zbraniecki — DISMISSED** [review]
thanks! looks great!

Optionally, you may want to fix (or comment) about the non-canonicalized fast-path on `und`.

File: components/locid/src/subtags/language.rs
Comment (robertbastian) [reply components/locid/src/subtags/language.rs:82]: I probably introduced this when I moved the check before the lowercasing. However after the merge this is deleted.

File: components/locid/src/parser/mod.rs
Comment (zbraniecki) [inline components/locid/src/parser/mod.rs:52]: `nit`: reverse the order (and consider `['-', '_'].contains` - optimize for one more lookup on the happy path.

File: components/locid/src/parser/mod.rs
Comment (zbraniecki) [inline components/locid/src/parser/mod.rs:59]: same.

**Review by zbraniecki — APPROVED** [review]
lgtm with one nit!

File: components/locid/src/parser/mod.rs
Comment (robertbastian) [reply components/locid/src/parser/mod.rs:52]: `contains` isn't `const`

**Review by sffc — APPROVED** [review]
My comments can be resolved in a follow-up if you prefer

**Pull request merged**
